File: app/src/main/java/com/example/myapplication/Adapter.java
package com.example.myapplication;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class Adapter extends RecyclerView.Adapter<Adapter.MyHolder> {

    private final String[] titles;
    private final int[] images;

    public Adapter(String[] titles, int[] images) {
        this.titles = titles;
        this.images = images;
    }

    @NonNull
    @Override
    public MyHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.data_layer, parent, false);
        return new MyHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MyHolder holder, int position) {
        holder.wallCoverTitle.setText(titles[position]);
        holder.wallCover.setImageResource(images[position]);
    }

    @Override
    public int getItemCount() {
        return titles.length;
    }

    static class MyHolder extends RecyclerView.ViewHolder {
        ImageView wallCover;
        TextView wallCoverTitle;

        public MyHolder(@NonNull View itemView) {
            super(itemView);
            wallCover = itemView.findViewById(R.id.wall_covers);
            wallCoverTitle = itemView.findViewById(R.id.wall_cover_titles);
        }
    }
}

File: app/src/main/java/com/example/myapplication/wallpaper/Adapter.kt
package com.example.myapplication.wallpaper

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import androidx.appcompat.app.AppCompatActivity
import com.example.myapplication.R


class Adapter(
    private val context: Context,
    private val wallpaperList: MutableList<Pair<String, Pair<Int, Class<out AppCompatActivity>>>>
) : RecyclerView.Adapter<Adapter.MyHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.data_layer, parent, false)
        return MyHolder(view)
    }

    override fun onBindViewHolder(holder: MyHolder, position: Int) {
        val (title, imageResIdAndActivity) = wallpaperList[position]
        val (imageResId, activityClass) = imageResIdAndActivity

        holder.wallCoverTitle.text = title

        // Load image using Glide
        Glide.with(holder.itemView.context)
            .load(imageResId)
            .placeholder(R.drawable.wallicon) // Placeholder while loading
            .error(R.drawable.baseline_front_hand_24) // Error image if loading fails
            .into(holder.wallCover)

        holder.itemView.setOnClickListener {
            try {
                val intent = Intent(context, activityClass)
                context.startActivity(intent)
            } catch (e: Exception) {
                e.printStackTrace()
                // Optionally show a toast or error message
            }
        }
    }

    override fun getItemCount(): Int = wallpaperList.size

    inner class MyHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val wallCover: ImageView = itemView.findViewById(R.id.wall_cover)
        val wallCoverTitle: TextView = itemView.findViewById(R.id.wall_cover_title)
    }
}
